README.md
# React DS
> Tiny (7KB) React Drag-to-Select component (with no dependencies! with support for touch devices!

[![Travis](https://img.shields.io/travis/aurbano/react-ds.svg)](https://travis-ci.org/aurbano/react-ds)
[![npm](https://img.shields.io/npm/v/react-ds.svg)](https://www.npmjs.com/package/react-ds)
[![Coverage Status](https://coveralls.io/repos/github/aurbano/react-ds/badge.svg?branch=master)](https://coveralls.io/github/aurbano/react-ds?branch=master)
[![npm](https://img.shields.io/npm/dm/react-ds.svg)](https://www.npmjs.com/package/react-ds)
[![npm](https://img.shields.io/npm/l/react-ds.svg)](https://www.npmjs.com/package/react-ds)
[![Codacy grade](https://img.shields.io/codacy/grade/e2589a609bdc4c56bd49c232a65dab4e.svg)](https://www.codacy.com/app/aurbano/react-ds)

[![react-ds gif](https://thumbs.gfycat.com/FatYellowKid-size_restricted.gif)](https://gfycat.com/gifs/detail/fatyellowkid)

I wrote this library because I couldn't find any existing one to do selections without having to wrap the items in their component.

In some cases you really need an unobtrusive way to make items selectable, this will do just that.


## Installation

```console
$ npm i react-ds
```
Or if you prefer yarn
```console
$ yarn add react-ds
```

## Usage

```jsx
import Selection from 'react-ds';

// target (ref) is the parent component (so that selects only happen when clicking and dragging on it)
// elements (refs[]) is an array of refs to the components that are selectable
<Selection
    target={ ref }
    elements={ refs[] }
    onSelectionChange={ this.handleSelection }
/>
```

### Props

#### `target`

Element where the selection should be applied to. This is to scope the mouse/touch event handlers and make sure that it doesn't affect your whole web app.

It must be a React `ref`, it should also exist, so you may want to check if it's already initialized before rendering the `Selection` component.

#### `elements`

Array of refs to the elements that are selectable. The `Selection` component will use this to get their location and sizes to determine whether they are within the selection area.

The should exist before rendering the `Selection` component.

#### `onSelectionChange`

Function that will be executed when the selection changes. An array of element indexes will be passed (with the same indexes as the `elements` prop).

This is where you want to update your state, to highlight them as selected for example.

#### `onHighlightChange`

While dragging, `onHighlightChange` is called only when the highlighted elements have changed.

When the mouse is released, it will be called with an empty array.

![onHighlightChange Example](https://user-images.githubusercontent.com/1640952/61724315-0269b280-ad6e-11e9-899c-4466e090cb13.gif)

#### `offset` *(Optional)*

This is used to calculate the coordinates of the mouse when drawing the Selection box, since the mouse events gives coordinates relative to the document, but the Selection box may have a different parent.

Essentially you need to pass the offset of the parent element where the Selection is being rendered. If it's rendered in the same component as the items to be selected then the default value will work fine.

If passing your own offset keep in mind that `getBoundingClientRect()` depends on the scroll, so you may want to do something like this:

```js
const boundingBox = target.getBoundingClientRect();
const offset = {
  top: boundingBox.top + window.scrollY,
  left: boundingBox.left + window.scrollX,
};
```

#### `style` *(Optional)*

If you want to override the styles for the selection area, you can either pass any styles here, or use css and declare any styles on the `.react-ds-border` class.

The styles are merged, so you can override just one property if you need (typically the `zIndex`).

The default styles are:

```js
const style = {
  position: 'absolute',
  background: 'rgba(159, 217, 255, 0.3)',
  border: 'solid 1px rgba(123, 123, 123, 0.61)',
  zIndex: 9,
  cursor: 'crosshair',
}
```

#### `ignoreTargets` *(Optional)*

Specify an array of CSS3 selectors for DOM targets that should be ignored when initiating a selection. i.e. `['div', 'div > p', '#someId']`

>This is specially useful because `react-ds` uses native browser events that bypass React's event queue, so you won't be able to `stopPropagation` as usual.

```jsx
<Selection
    target={ this.state.ref}
    elements={ this.state.elRefs }
    onSelectionChange={ this.handleSelection }
    ignoreTargets={ ['.handle'] }
/>
```

## Example

This example was taken from [`example/app/src/Example.js`](https://github.com/aurbano/react-ds/blob/master/example/app/src/Example.js) which you can see running at https://aurbano.eu/react-ds/

```jsx
import React from 'react';
import PropTypes from 'prop-types';
import Selection from 'react-ds';

export default class Example extends React.PureComponent {

  constructor() {
    super();

    this.state = {
      ref: null,
      elRefs: [],
      selectedElements: [], // track the elements that are selected
    };
  }

  handleSelection = (indexes) => {
    this.setState({
      selectedElements: indexes,
    });
  };

  getStyle = (index) => {
    if (this.state.selectedElements.indexOf(index) > -1) {
      // Selected state
      return {
        background: '#2185d0',
        borderColor: '#2185d0',
        color: 'white',
      };
    }
    return {};
  };

  addElementRef = (ref) => {
    const elRefs = this.state.elRefs;
    elRefs.push(ref);
    this.setState({
      elRefs,
    });
  };

  renderSelection() {
    if (!this.state.ref || !this.state.elRefs) {
      return null;
    }
    return (
      <Selection
        target={ this.state.ref}
        elements={ this.state.elRefs }
        onSelectionChange={ this.handleSelection }
        style={ this.props.style }
      />
    );
  }

  render() {
    const selectableElements = [
      'one',
      'another',
      'hey there',
      'item',
      'two',
      'three',
      'something longer?',
      'last'
    ];
    return (
      <div ref={ (ref) => { this.setState({ ref }); } } className='item-container'>
        { selectableElements.map((el, index) => (
          <div
            key={ el }
            ref={ this.addElementRef }
            style={ this.getStyle(index) }
            className='item'
          >
            { el }
          </div>
        )) }
        { this.renderSelection() }
      </div>
    );
  }
}

Example.PropTypes = {
  style: PropTypes.object,
};
```

## Contributing

Only edit the files in the `src` folder. I'll update `dist` manually before publishing new versions to npm.

To run the tests simply run `npm test`. Add tests as you see fit to the `test` folder, they must be called `{string}.test.js`.

## Meta

Copyright &copy; [Alejandro U. Alvarez](https:/aurbano.eu) 2017. MIT Licensed.


CHANGELOG.md
# Change Log

> I haven't been maintaining this changelog.
>
> Check out the [releases](https://github.com/aurbano/react-ds/releases) section instead.

------

# Old Changelog

## [v1.7.0](https://github.com/aurbano/react-ds/tree/v1.7.0) (2017-07-27)
[Full Changelog](https://github.com/aurbano/react-ds/compare/v1.6.0...v1.7.0)

**Closed issues:**

- Add option to ignore certain targets when initiating selects [\#7](https://github.com/aurbano/react-ds/issues/7)

## [v1.6.0](https://github.com/aurbano/react-ds/tree/v1.6.0) (2017-07-25)
[Full Changelog](https://github.com/aurbano/react-ds/compare/v1.5.0...v1.6.0)

## [v1.5.0](https://github.com/aurbano/react-ds/tree/v1.5.0) (2017-07-13)
[Full Changelog](https://github.com/aurbano/react-ds/compare/v1.4.1...v1.5.0)

## [v1.4.1](https://github.com/aurbano/react-ds/tree/v1.4.1) (2017-07-13)
[Full Changelog](https://github.com/aurbano/react-ds/compare/v1.4.0...v1.4.1)

**Closed issues:**

- Move react to peerDependencies [\#5](https://github.com/aurbano/react-ds/issues/5)

## [v1.4.0](https://github.com/aurbano/react-ds/tree/v1.4.0) (2017-07-12)
[Full Changelog](https://github.com/aurbano/react-ds/compare/v1.3.0...v1.4.0)

**Fixed bugs:**

- Items not properly selected when scrolled [\#4](https://github.com/aurbano/react-ds/issues/4)

## [v1.3.0](https://github.com/aurbano/react-ds/tree/v1.3.0) (2017-07-12)
[Full Changelog](https://github.com/aurbano/react-ds/compare/v1.2.0...v1.3.0)

## [v1.2.0](https://github.com/aurbano/react-ds/tree/v1.2.0) (2017-07-12)
[Full Changelog](https://github.com/aurbano/react-ds/compare/v1.1.0...v1.2.0)

**Implemented enhancements:**

- Add examples [\#1](https://github.com/aurbano/react-ds/issues/1)

## [v1.1.0](https://github.com/aurbano/react-ds/tree/v1.1.0) (2017-07-12)
[Full Changelog](https://github.com/aurbano/react-ds/compare/v1.0.3...v1.1.0)

## [v1.0.3](https://github.com/aurbano/react-ds/tree/v1.0.3) (2017-07-12)
[Full Changelog](https://github.com/aurbano/react-ds/compare/v1.0.2...v1.0.3)

**Implemented enhancements:**

- Setup CI [\#3](https://github.com/aurbano/react-ds/issues/3)

## [v1.0.2](https://github.com/aurbano/react-ds/tree/v1.0.2) (2017-07-12)
[Full Changelog](https://github.com/aurbano/react-ds/compare/v1.0.1...v1.0.2)

**Implemented enhancements:**

- Generate dist folder with babel without flow [\#2](https://github.com/aurbano/react-ds/issues/2)

## [v1.0.1](https://github.com/aurbano/react-ds/tree/v1.0.1) (2017-07-12)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


src/index.js
// @flow

import React from 'react';
import PropTypes from 'prop-types';

export type Point = {
  x: number,
  y: number,
}

export type Box = {
  left: number,
  top: number,
  width: number,
  height: number,
}

type Props = {
  disabled?: boolean,
  target: HTMLElement,
  onSelectionChange(elements: Array<any>): void,
  onHighlightChange(elements: Array<any>): void,
  elements: Array<HTMLElement>,
  // eslint-disable-next-line react/no-unused-prop-types
  offset?: {
    // eslint-disable-next-line react/no-unused-prop-types
    top: number,
    // eslint-disable-next-line react/no-unused-prop-types
    left: number,
  },
  style?: any,
  zoom?: number,
  ignoreTargets?: Array<string>,
};

type State = {
  mouseDown: boolean,
  startPoint: ?Point,
  endPoint: ?Point,
  selectionBox: ?Box,
  offset: {
    top: number,
    left: number,
  },
};

function getOffset(props: Props) {
  let offset = {
    top: 0,
    left: 0,
  };
  if (props.offset) {
    offset = {
      ...props.offset,
    };
  } else if (props.target) {
    const boundingBox = props.target.getBoundingClientRect();
    offset.top = boundingBox.top + window.scrollY;
    offset.left = boundingBox.left + window.scrollX;
  }
  return offset;
}

export default class Selection extends React.PureComponent<Props, State> { // eslint-disable-line react/prefer-stateless-function
  props: Props;
  state: State;
  selectedChildren: Array<number>;
  highlightedChildren: Array<number>;

  constructor(props: Props) {
    super(props);

    this.state = {
      mouseDown: false,
      startPoint: null,
      endPoint: null,
      selectionBox: null,
      offset: getOffset(props),
    };

    this.selectedChildren = [];
    this.highlightedChildren = [];
  }

  componentDidMount() {
    this.reset();
    this.bind();
  }

  componentWillReceiveProps(nextProps: Props) {
    this.setState({
      offset: getOffset(nextProps),
    });
  }

  componentDidUpdate() {
    this.reset();
    this.bind();
    if (this.state.mouseDown && this.state.selectionBox) {
      this.updateCollidingChildren(this.state.selectionBox);
    }
  }

  componentWillUnmount() {
    this.reset();
    window.document.removeEventListener('mousemove', this.onMouseMove);
    window.document.removeEventListener('mouseup', this.onMouseUp);
  }

  bind = () => {
    this.props.target.addEventListener('mousedown', this.onMouseDown);
    this.props.target.addEventListener('touchstart', this.onTouchStart);
  };

  reset = () => {
    if (this.props.target) {
      this.props.target.removeEventListener('mousedown', this.onMouseDown);
    }
  };

  init = (e: Event, x: number, y: number): boolean => {
    if (this.props.ignoreTargets) {
      const Target = (e.target: any);
      if (!Target.matches) {
        // polyfill matches
        const defaultMatches = (s: string) => (
          [].indexOf.call(window.document.querySelectorAll(s), this) !== -1
        );
        Target.matches =
          Target.matchesSelector ||
          Target.mozMatchesSelector ||
          Target.msMatchesSelector ||
          Target.oMatchesSelector ||
          Target.webkitMatchesSelector ||
          defaultMatches;
      }
      if (Target.matches && Target.matches(this.props.ignoreTargets.join(','))) {
        return false;
      }
    }
    const nextState = {};

    const zoom = this.props.zoom || 1;

    nextState.mouseDown = true;
    nextState.startPoint = {
      x: (x * zoom) - this.state.offset.left,
      y: (y * zoom) - this.state.offset.top,
    };

    this.setState(nextState);
    return true;
  };

  /**
   * On root element mouse down
   * The event should be a MouseEvent | TouchEvent, but flow won't get it...
   * @private
   */
  onMouseDown = (e: MouseEvent | any) => {
    if (this.props.disabled || e.button === 2 || (e.nativeEvent && e.nativeEvent.which === 2)) {
      return;
    }

    if (this.init(e, e.pageX, e.pageY)) {
      window.document.addEventListener('mousemove', this.onMouseMove);
      window.document.addEventListener('mouseup', this.onMouseUp);
    }
  };

  onTouchStart = (e: TouchEvent) => {
    if (this.props.disabled || !e.touches || !e.touches[0] || e.touches.length > 1) {
      return;
    }

    if (this.init(e, e.touches[0].pageX, e.touches[0].pageY)) {
      window.document.addEventListener('touchmove', this.onTouchMove);
      window.document.addEventListener('touchend', this.onMouseUp);
    }
  };

  /**
   * On document element mouse up
   * @private
   */
  onMouseUp = () => {
    window.document.removeEventListener('touchmove', this.onTouchMove);
    window.document.removeEventListener('mousemove', this.onMouseMove);
    window.document.removeEventListener('mouseup', this.onMouseUp);
    window.document.removeEventListener('touchend', this.onMouseUp);

    this.setState({
      mouseDown: false,
      startPoint: null,
      endPoint: null,
      selectionBox: null,
    });

    this.props.onSelectionChange(this.selectedChildren);
    if (this.props.onHighlightChange) {
      this.highlightedChildren = [];
      this.props.onHighlightChange(this.highlightedChildren);
    }
    this.selectedChildren = [];
  };

  /**
   * On document element mouse move
   * @private
   */
  onMouseMove = (e: MouseEvent) => {
    e.preventDefault();
    if (this.state.mouseDown) {
      const endPoint: Point = {
        x: e.pageX - this.state.offset.left,
        y: e.pageY - this.state.offset.top,
      };

      this.setState({
        endPoint,
        selectionBox: this.calculateSelectionBox(
          this.state.startPoint,
          endPoint
        ),
      });
    }
  };

  onTouchMove = (e: TouchEvent) => {
    e.preventDefault();
    if (this.state.mouseDown) {
      const endPoint: Point = {
        x: e.touches[0].pageX - this.state.offset.left,
        y: e.touches[0].pageY - this.state.offset.top,
      };

      this.setState({
        endPoint,
        selectionBox: this.calculateSelectionBox(
          this.state.startPoint,
          endPoint
        ),
      });
    }
  };

  /**
   * Calculate if two segments overlap in 1D
   * @param lineA [min, max]
   * @param lineB [min, max]
   */
  lineIntersects = (lineA: [number, number], lineB: [number, number]): boolean => (
    lineA[1] >= lineB[0] && lineB[1] >= lineA[0]
  );

  /**
   * Detect 2D box intersection - the two boxes will intersect
   * if their projections to both axis overlap
   * @private
   */
  boxIntersects = (boxA: Box, boxB: Box): boolean => {
    // calculate coordinates of all points
    const boxAProjection = {
      x: [boxA.left, boxA.left + boxA.width],
      y: [boxA.top, boxA.top + boxA.height],
    };

    const boxBProjection = {
      x: [boxB.left, boxB.left + boxB.width],
      y: [boxB.top, boxB.top + boxB.height],
    };

    return this.lineIntersects(boxAProjection.x, boxBProjection.x) &&
           this.lineIntersects(boxAProjection.y, boxBProjection.y);
  };

  /**
   * Updates the selected items based on the
   * collisions with selectionBox,
   * also updates the highlighted items if they have changed
   * @private
   */
  updateCollidingChildren = (selectionBox: Box) => {
    this.selectedChildren = [];
    if (this.props.elements) {
      this.props.elements.forEach((ref, $index) => {
        if (ref) {
          const refBox = ref.getBoundingClientRect();
          const tmpBox = {
            top: (refBox.top - this.state.offset.top) + window.scrollY,
            left: (refBox.left - this.state.offset.left) + window.scrollX,
            width: ref.clientWidth,
            height: ref.clientHeight,
          };

          if (this.boxIntersects(selectionBox, tmpBox)) {
            this.selectedChildren.push($index);
          }
        }
      });
    }
    if (this.props.onHighlightChange && JSON.stringify(this.highlightedChildren) !== JSON.stringify(this.selectedChildren)) {
      this.highlightedChildren = [...this.selectedChildren];
      if (window.requestAnimationFrame) {
        window.requestAnimationFrame(() => {
          this.props.onHighlightChange(this.highlightedChildren);
        });
      } else {
        this.props.onHighlightChange(this.highlightedChildren);
      }
    }
  };

  /**
   * Calculate selection box dimensions
   * @private
   */
  calculateSelectionBox = (startPoint: ?Point, endPoint: ?Point) => {
    if (!this.state.mouseDown || !startPoint || !endPoint) {
      return null;
    }

    // The extra 1 pixel is to ensure that the mouse is on top
    // of the selection box and avoids triggering clicks on the target.
    const left = Math.min(startPoint.x, endPoint.x) - 1;
    const top = Math.min(startPoint.y, endPoint.y) - 1;
    const width = Math.abs(startPoint.x - endPoint.x) + 1;
    const height = Math.abs(startPoint.y - endPoint.y) + 1;

    return {
      left,
      top,
      width,
      height,
    };
  };

  /**
   * Render
   */
  render() {
    const style = {
      position: 'absolute',
      background: 'rgba(159, 217, 255, 0.3)',
      border: 'solid 1px rgba(123, 123, 123, 0.61)',
      zIndex: 9,
      cursor: 'crosshair',
      ...this.state.selectionBox,
      ...this.props.style,
    };
    if (!this.state.mouseDown || !this.state.endPoint || !this.state.startPoint) {
      return null;
    }
    return (
      <div className='react-ds-border' style={ style } />
    );
  }
}

Selection.propTypes = {
  target: PropTypes.object,
  disabled: PropTypes.bool,
  onSelectionChange: PropTypes.func.isRequired,
  onHighlightChange: PropTypes.func,
  elements: PropTypes.array.isRequired,
  // eslint-disable-next-line react/no-unused-prop-types
  offset: PropTypes.object,
  zoom: PropTypes.number,
  style: PropTypes.object,
  ignoreTargets: PropTypes.array,
};
